### PR TITLE
PLT-6581 'runIndexers' retries connecting to node if node is stopped

### PR DESCRIPTION
**Pull request by koslambrou**
Here is the log output:

```
[marconi-sidechain:Warning:13] [2023-08-11 00:36:28.21 UTC] Could not connect to Cardano node: socket file /home/kolam/cardano-node/preprod/cardano-node.socket does not exist. Retrying in 30s ...
[marconi-sidechain:Info:13] [2023-08-11 00:37:01.82 UTC] Resumable points for each indexer:
    [ ChainPoint(Slot 25624547, BlockHash db6f02a548d868462650feb0c457bd0e038269da3699c77ab668419898ac04b7)
    , ChainPoint(Slot 25574407, BlockHash 673fa32353ea5021cb0e3a694d4bfaee201ec9c83118f53f511936fd36fb40b0)
    , ChainPoint(Slot 25624589, BlockHash 45db006dd75a2ef0c5600e43b9bad0558bb5ee8e45951ceb8f2558c54c4d2741) ]
[marconi-sidechain:Info:13] [2023-08-11 00:37:01.83 UTC] Starting from ChainPoint(Slot 25574407, BlockHash 673fa32353ea5021cb0e3a694d4bfaee201ec9c83118f53f511936fd36fb40b0). 
[marconi-sidechain:Info:13] [2023-08-11 00:37:11.90 UTC] Synchronising (71.14%). Current synced point is ChainPoint(Slot 25632899, BlockHash 1d3cca44232eaa1e99e8f1d2789b23f0e3af1c270af35a1e6ad4f02fea31aaa4) and current node tip is ChainTip(Slot 36030999, BlockHash e59fcaf9a4ae9592b5eef4b89c07489112622a41adf9ef3
120c5360bc76e910e, BlockNo 1257955). Processed 2553 blocks and 1 rollbacks in the last 10s (253 blocks/s).
[marconi-sidechain:Info:13] [2023-08-11 00:37:21.97 UTC] Synchronising (71.22%). Current synced point is ChainPoint(Slot 25662115, BlockHash 1f1d8988fb02a9675525cf337b15cde86534c61d6629d5adfd3071b94d663d34) and current node tip is ChainTip(Slot 36030999, BlockHash e59fcaf9a4ae9592b5eef4b89c07489112622a41adf9ef3
120c5360bc76e910e, BlockNo 1257955). Processed 1311 blocks and 0 rollbacks in the last 10s (130 blocks/s).
[marconi-sidechain:Info:13] [2023-08-11 00:37:24.25 UTC] Stopping indexing. Waiting for indexers to finish their work (timeout after 180s) ...
[marconi-sidechain:Info:13] [2023-08-11 00:37:24.26 UTC] Done!
[marconi-sidechain:Warning:13] [2023-08-11 00:37:24.26 UTC] Could not connect to Cardano node: socket file /home/kolam/cardano-node/preprod/cardano-node.socket does not exist. Retrying in 30s ...
[marconi-sidechain:Warning:13] [2023-08-11 00:37:54.29 UTC] Could not connect to Cardano node: socket file /home/kolam/cardano-node/preprod/cardano-node.socket does not exist. Retrying in 30s ...
[marconi-sidechain:Info:13] [2023-08-11 00:38:27.54 UTC] Resumable points for each indexer:
    [ ChainPoint(Slot 25620682, BlockHash 485310044d4a29b976c6671858e93335da86c9b379450ee5e1086f85bc1c1ab2)
    , ChainPoint(Slot 25574407, BlockHash 673fa32353ea5021cb0e3a694d4bfaee201ec9c83118f53f511936fd36fb40b0)
    , ChainPoint(Slot 25620682, BlockHash 485310044d4a29b976c6671858e93335da86c9b379450ee5e1086f85bc1c1ab2) ]
[marconi-sidechain:Info:13] [2023-08-11 00:38:27.55 UTC] Starting from ChainPoint(Slot 25574407, BlockHash 673fa32353ea5021cb0e3a694d4bfaee201ec9c83118f53f511936fd36fb40b0). 
^C[marconi-sidechain:Info:13] [2023-08-11 00:38:31.01 UTC] Stopping indexing. Waiting for indexers to finish their work (timeout after 180s) ...
[marconi-sidechain:Info:13] [2023-08-11 00:38:31.33 UTC] Done!
```

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting main unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [ ] Reviewer requested

### CODE REVIEW

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs [hunk 1]
```diff
@@ -56,27 +62,81 @@ runIndexer
...
-    (readEvent eventQueue cBox)
+  withTrace c traceName $ \trace -> runChainSyncClientWithRetries trace
+  where
+    secondsToWait :: Int
+    secondsToWait = 30
+
+    runChainSyncClientWithRetries
+      :: Trace IO Text
+      -> IO ()
+    runChainSyncClientWithRetries trace = do
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs:74]: A classical strategy for retry is to start with a quick retry and to space them more and more. We can either implement such strategy now or do it in another ticket
We could use something like:

```
data RetryState =
  RetryState
  { secondsBeforeNextRetry :: Word
  , initialWait :: Word
  , maxWait :: Word
  }
```

We pass this type to `runChainSyncClientWithRetries` (or we do use a state monad), erinitialise to `initialWait` on success, double the time on each retry.

Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs:74]: Doesn't seem too complicated. I can do it as part of this PR.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs [hunk 2]
```diff
@@ -56,27 +62,81 @@ runIndexer
   -> indexer (WithDistance BlockEvent)
   -> IO ()
 runIndexer socketPath networkId _startingPoint traceName indexer = do
-  securityParam <- Utils.toException $ Utils.querySecurityParam @Void networkId socketPath
-  eventQueue <- STM.newTBQueueIO $ fromIntegral securityParam
-  cBox <- Concurrent.newMVar indexer
   c <- Tracing.defaultConfigStdout
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs:65]: While we're here, change the variable name?

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs [hunk 3]
```diff
@@ -56,27 +62,81 @@ runIndexer
...
+
+    runChainSyncClientWithRetries
+      :: Trace IO Text
+      -> IO ()
+    runChainSyncClientWithRetries trace = do
+      catches
+        (runChainSyncClient trace)
+        [ Handler $
+            \case
+              e | isDoesNotExistError e -> handleRunIndexersException trace
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs:79]: Can you put a comment to explain why we catch this one? I don't get it.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs [hunk 4]
```diff
@@ -56,27 +62,81 @@ runIndexer
...
+      runChainSyncClientWithRetries trace
+
+    runChainSyncClient
+      :: Trace IO Text
+      -> IO ()
+    runChainSyncClient trace = do
+      securityParam <- Utils.toException $ Utils.querySecurityParam @Void networkId socketPath
+      eventQueue <- STM.newTBQueueIO $ fromIntegral securityParam
+      cBox <- Concurrent.newMVar indexer
+      let io =
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs:109]: Rename `io`? `streamEvents` maybe?

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs [hunk 5]
```diff
@@ -56,27 +62,81 @@ runIndexer
...
+            withChainSyncEventEpochNoStream
+              socketPath
+              networkId
+              [Core.genesis]
+              (mkEventStream eventQueue . chainSyncEventStreamLogging trace)
+          handleException NoIntersectionFound =
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs:115]: Same, `handleException` may be `whenNoIntersectionFound`?

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs [hunk 6]
```diff
@@ -56,27 +62,81 @@ runIndexer
...
+            \case
+              e | isDoesNotExistError e -> handleRunIndexersException trace
+              e -> throwIO e
+        , Handler $
+            \case
+              MuxError (MuxIOException _) _ -> handleRunIndexersException trace
+              e -> throwIO e
+        ]
+
+    handleRunIndexersException
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs:87]: shouldn't it be `handleChainSyncClientException`?

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs [hunk 7]
```diff
@@ -659,68 +663,108 @@ runIndexers
...
+    waitForIndexersToFinishProcessingLastEvent trace coordinator = do
+      let secondsBeforeTimeout = 180
+      logInfo trace $
+        "Stopping indexing. Waiting for indexers to finish their work (timeout after "
+          <> Text.pack (show secondsBeforeTimeout)
+          <> "s) ..."
+      void $
+        timeout (secondsBeforeTimeout * 1_000_000) $
+          waitQSemN (coordinator ^. barrier) (coordinator ^. indexerCount)
+      logInfo trace "Done!"
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs:767]: Reapply the same name chaes as in the experimental runner, if we still care about the quality of this part of the codebase.

**Review by berewt — APPROVED** [review]
A few minor changes, mostly name changes

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs [hunk 8]
```diff
@@ -56,27 +62,81 @@ runIndexer
   -> indexer (WithDistance BlockEvent)
   -> IO ()
 runIndexer socketPath networkId _startingPoint traceName indexer = do
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs:64]: As we have explicit export now in the module, maybe we can avoid putting everything in the where clause, and try to split it into smaller groups of cohesive functions, to improve readability. Even with small functions, a where clause of 80 lines is big.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs (hunk 4 above)
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs]: Changed to `runChainSyncStream`

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs (hunk 5 above)
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs]: Changed it.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs (hunk 6 above)
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs]: Changed to `handleCantConnectToNodeException`.

File: marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs (hunk 7 above)
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Indexers.hs]: Applied all comments for the experimental runner as well.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs (hunk 8 above)
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs]: Moved out the `readEvent` function.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs (hunk 1 above)
Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs]: The change is done. However, if the node connection is re-established, then we increment the number of seconds before retry based on the previous number. Not sure how to make this better.

Here's the new log:

```
[marconi-chain-index:Warning:11] [2023-08-14 17:02:37.15 UTC] Could not connect to Cardano node: socket file /home/kolam/cardano-node/preprod/cardano-node.socket does not exist. Retrying in 30s ...
[marconi-chain-index:Warning:11] [2023-08-14 17:03:07.18 UTC] Could not connect to Cardano node: socket file /home/kolam/cardano-node/preprod/cardano-node.socket does not exist. Retrying in 60s ...
[marconi-chain-index:Info:11] [2023-08-14 17:04:07.33 UTC] Resumable points for each indexer:
    [ ChainPoint(Slot 993935, BlockHash ffb2e3f74984692541a11301c8e5851be8db42fcd65230f049814824c700bd20)
    , ChainPoint(Slot 993935, BlockHash ffb2e3f74984692541a11301c8e5851be8db42fcd65230f049814824c700bd20)
    , ChainPoint(Slot 993935, BlockHash ffb2e3f74984692541a11301c8e5851be8db42fcd65230f049814824c700bd20)
    , ChainPoint(Slot 993935, BlockHash ffb2e3f74984692541a11301c8e5851be8db42fcd65230f049814824c700bd20) ]
[marconi-chain-index:Info:11] [2023-08-14 17:04:07.34 UTC] Starting from ChainPoint(Slot 993935, BlockHash ffb2e3f74984692541a11301c8e5851be8db42fcd65230f049814824c700bd20). 
[marconi-chain-index:Info:11] [2023-08-14 17:04:17.34 UTC] Synchronising (4.44%). Current synced point is ChainPoint(Slot 1614855, BlockHash 770d8355c9958b9cebc210e95c6d209bec09020788f98b87f7c3ecd9cde10e41) and current node tip is ChainTip(Slot 36349430, BlockHash 83cd3e7e437bc7ea9a9446455356786b158934c55ef90fe
b3fdd57a6fb6cee6c, BlockNo 1270014). Processed 31103 blocks and 1 rollbacks in the last 10s (3110 blocks/s).
[marconi-chain-index:Info:11] [2023-08-14 17:04:27.35 UTC] Synchronising (5.95%). Current synced point is ChainPoint(Slot 2162421, BlockHash 2f7107e6d00f4a4867b8525d31150f45180c033b285c93564a150e6974756cdb) and current node tip is ChainTip(Slot 36349430, BlockHash 83cd3e7e437bc7ea9a9446455356786b158934c55ef90fe
b3fdd57a6fb6cee6c, BlockNo 1270014). Processed 27247 blocks and 0 rollbacks in the last 10s (2721 blocks/s).
[marconi-chain-index:Info:11] [2023-08-14 17:04:37.35 UTC] Synchronising (7.44%). Current synced point is ChainPoint(Slot 2703595, BlockHash 582dda8c9997158e94a631c5d395c971c448580319c45f3bbb58f7d04d8e0fe8) and current node tip is ChainTip(Slot 36349430, BlockHash 83cd3e7e437bc7ea9a9446455356786b158934c55ef90fe
b3fdd57a6fb6cee6c, BlockNo 1270014). Processed 26845 blocks and 0 rollbacks in the last 10s (2684 blocks/s).
[marconi-chain-index:Info:11] [2023-08-14 17:04:39.35 UTC] Stopping indexing. Waiting for indexers to finish their work (timeout after 180s) ...
[marconi-chain-index:Info:11] [2023-08-14 17:04:39.35 UTC] Done!
[marconi-chain-index:Warning:11] [2023-08-14 17:04:39.35 UTC] Could not connect to Cardano node: socket file /home/kolam/cardano-node/preprod/cardano-node.socket does not exist. Retrying in 120s ...
```

See the last retry log.

File: marconi-chain-index/src/Marconi/ChainIndex/Node/Client/Retry.hs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+ )
+import Prettyprinter.Render.Text (renderStrict)
+import System.Exit (exitFailure)
+import System.IO.Error (isDoesNotExistError)
+
+data RetryState = RetryState
+  { totalWaitTime :: !Int
+  , secondsBeforeNextRetry :: !Int
+  , nextRetryIncrement :: Int -> Int
+  , maxWaitTime :: !Int
```
Comment (berewt) [inline marconi-chain-index/src/Marconi/ChainIndex/Node/Client/Retry.hs:32]: Can we use `Maybe Int` to retry indefinitely?

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs (hunk 1 above)
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs]: We could probably introduce a `baseTime` and revert to it on success. I'm more a afraid of a potential memory leak with the current version, as I'm not sure the recursive calls to retry can be optimised.

Comment (koslambrou) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs]: > We could probably introduce a baseTime and revert to it on success.

The problem is that we don't know what "on success" is. The function just runs indefinitely. Unless we consider that the function is "successful" if it has run after a certain amount of time, but that seems arbitrary.

> I'm more a afraid of a potential memory leak with the current version, as I'm not sure the recursive calls to retry can be optimised.

Good point. Do you have any enhancement ideas?

Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs]: Sorry, I didn't take the time to think about it. So no, nothing at the moment. We can look at it when I'll be back or you can monitor it and check if it has a memory impact. In the current state, an infinite and linear retry time might be easier/more stable, but I let you decide.

**Pull request merged**